spec/registry.md

```diff
@@ -622,14 +622,24 @@ All other values produce an _Invalid Expression_ error.
 ### Non-Negative Integer Options
 
 Some _options_ of number _functions_ are defined to take a "non-negative integer".
-In most cases, these control the number of various kinds of digits.
-
-A "non-negative integer" is one of:
-- an implementation-defined integer numeric type where the value of the 
-  encoded number is not less than zero
-- a literal consisting only of ASCII digits U+0030 through U+0039;
-  An implementation MAY define an upper limit on the length of such a literal
-  or of the value encoded by such a literal.
+In this specification, these _options_ are used to control aspects of numeric display
+such as the number of fraction, integer, or significant digits.
+
+A "non-negative integer" is a variable reference or literal whose value evaluates 
+as an integer greater than or equal to zero.
+If the value is passed as a literal
+it MUST contain only the ASCII digits U+0030 through U+0039.
```

Comment:
This is still problematic. Consider these two example messages:
```
literal: {42 :number minimumFractionDigits=2}
```
```
variable: {42 :number minimumFractionDigits=$foo}
```

With the current proposed language, formatting the first is valid, because there the `2` is a literal, but formatting the second with `{ foo: '2' }` as input is invalid, because it's a variable, and we're here establishing stricter rules for variables than literals.

The problem with that is that in order to make that differentiation, one of two things must happen:
1. The messageformat runtime implementation must parse the numeric value out of the literal before passing it to the `:number` implementation as an integer, or emitting an error if it doesn't match the expectations. This requires for `:number` to somehow communicate to the runtime its expectations on the types of its options, and for types to be introduced to the runtime. I really doubt that we want this.
2. The `:number` implementation must be able to tell where an option value comes from, so that it can apply one set of rules to literal values and another to variable values. In practice, with the examples as above, it would mean that a `:number` implementation would need to be called with
   ```
   number(context, Literal('42'), { minimumFractionDigits: Literal('2') })
   ```
   for the first message, and with
   ```
   number(context, Literal('42'), { minimumFractionDigits: Variable('2') })
   ```
   for the second, so that `Literal('2')` and `Variable('2')` could be differentiated.

   This would be a new, additional requirement on implementations, as without this addition it is possible for a valid runtime to call a `:number` implementation for either message as
   ```
   number(context, '42', { minimumFractionDigits: '2' })
   ```

So let's not introduce language here or elsewhere that differentiates the treatment of literal vs variable values, because that would add complexity to all custom functions in all implementations. Instead, let's just allow for a variable with a value of `'2'` to get treated the exact same as a literal `2` in options.

Reply:
The idea here is to impose implementation details on implementations of `:number` (and `:integer`), not on MF. There has to be a clear delineation for message authors about what literal values are accepted. I will try to clarify that.

Reply:
(also, I didn't read "passed as a literal" to only mean in the syntax, but also as the value of a variable)

Reply:
I still think there's some confusion because this text isn't defining a syntactic property, it's defining the runtime interface with a function implemented in another language, so saying "must accept a string" doesn't mean you can't write `maximumFractionDigits=|11|` in the syntax. Passing in a string of length 4 that begins and ends with a '|' should be an error if an integer option is expected. _Function implementations_ take parsed values (though of course this area is muddy because we avoid saying what a parsed value is.)

Reply:
> so saying "must accept a string" doesn't mean you can't write maximumFractionDigits=|11| in the syntax. Passing in a string of length 4 that begins and ends with a '|' should be an error if an integer option is expected

Nah. That thing to the right of the equals sign can't be anything _but_ a literal (or a variable with its `$` sigil). That means it can be quoted. We very explicitly say that there is no difference between `|11|` and `11`. As far as MF is concerned, all there is to the right of the equals is a literal. It is up to the number function to decide what to do with it.

A _literal_ of length 4 that begins and ends with a `|` looks like `maxiumumFractionDigits=|\|11\||` in our syntax.

Functions implementations can get typed values via variable references, e.g. `maximumFractionDigits=$maxDigits` where `$maxDigits` is some kind of `int`-like type. Otherwise it must parse a literal. This is "number functions" defining what literals it accepts and what it does with them.

Reply:
Maybe what's confusing is that there are two different parsers:

- The MessageFormat parser takes a MessageFormat string and produces a data model, which includes nodes representing _literals_ in the syntax.
- Each function implementation has its own parser that takes strings (or other implementation-defined types) and produces some implementation-defined type, in this case a number.

The first parser is responsible for stripping the '|' characters, so that the second parser doesn't need to have its own partial implementation of MessageFormat syntax. That's what I meant.

Reply:
I see what you mean. The registry spec is written from the outside in: it talks about what you can have in the _message_ syntax and uses MF's terminology for that. So the first parser is what is in play. You're correct that the number function implementation never sees the `|` in your example. It gets the literal string `11` and parses it. That implementation, by the way, is probably glue code that calls the ultimate formatter.

> that means defining common types across the interface

I don't agree. I don't think the **_default_** registry should touch types nor does it need to. The **_default_** registry should be written in a type-neutral way, allowing local implementations to define variable types (for operands or option values) as needed/expected by users. An implementation of the number functions in the registry (as an example) could be written entirely in a language that has no number type or integer type and could work entirely in terms of strings, with only strings/literals as input. (That would be weird, but it could happen). 

What `registry.md` defines is what the function is supposed to do; what operands are accepted (which mostly means: is there an accepted literal form?); and what options/option values are accepted (with the values defining what literal forms can be passed--if any--for implementation-defined variable values).

Reply:
I don't think the default registry can be completely type-neutral; it already refers to string types. That's what I mean by "a common type": I can't see how I would rewrite this spec without assuming the implementation had a string type. So why not talk about strings here, given that the formatting spec already says "The resolution of a text or literal MUST resolve to a string"?

Reply:
I turned the text hear upside down and I think it works better: now we presume a decimal integer string and permit other types. See what you think.